pkg/vault/contracts/VaultFactory.sol

```diff
@@ -64,28 +74,67 @@ contract VaultFactory is Authentication {
      * @param targetAddress Expected Vault address. The function will revert if the given salt does not deploy the
      * Vault to the target address.
      */
-    function create(bytes32 salt, address targetAddress) external authenticate {
-        if (isDisabled) {
-            revert VaultAlreadyCreated();
+    function create(
+        bytes32 salt,
+        address targetAddress,
+        bytes memory vaultCreationCode,
+        bytes memory vaultAdminCreationCode,
+        bytes memory vaultExtensionCreationCode
+    ) external authenticate {
+        if (vaultCreationCodeHash != keccak256(vaultCreationCode)) {
+            revert InvalidBytecode("Vault");
+        } else if (vaultAdminCreationCodeHash != keccak256(vaultAdminCreationCode)) {
+            revert InvalidBytecode("VaultAdmin");
+        } else if (vaultExtensionCreationCodeHash != keccak256(vaultExtensionCreationCode)) {
+            revert InvalidBytecode("VaultExtension");
         }
-        isDisabled = true;
 
         address vaultAddress = getDeploymentAddress(salt);
         if (targetAddress != vaultAddress) {
             revert VaultAddressMismatch();
         }
 
-        VaultAdmin vaultAdmin = new VaultAdmin(
-            IVault(vaultAddress),
-            _pauseWindowDuration,
-            _bufferPeriodDuration,
-            _minTradeAmount,
-            _minWrapAmount
+        VaultAdmin vaultAdmin = VaultAdmin(
+            payable(
+                Create2.deploy(
+                    0,
+                    bytes32(0x00),
+                    abi.encodePacked(
+                        vaultAdminCreationCode,
+                        abi.encode(
+                            IVault(vaultAddress),
+                            _pauseWindowDuration,
+                            _bufferPeriodDuration,
+                            _minTradeAmount,
+                            _minWrapAmount
+                        )
+                    )
+                )
+            )
+        );
+
+        require(
+            vaultExtensionCreationCodeHash == keccak256(vaultExtensionCreationCode),
+            "VaultFactory: invalid vaultExtensionCreationCode"
         );
-        VaultExtension vaultExtension = new VaultExtension(IVault(vaultAddress), vaultAdmin);
+        VaultExtension vaultExtension = VaultExtension(
+            payable(
+                Create2.deploy(
+                    0,
+                    bytes32(uint256(0x01)),
+                    abi.encodePacked(vaultExtensionCreationCode, abi.encode(vaultAddress, vaultAdmin))
+                )
+            )
+        );
+
         ProtocolFeeController feeController = new ProtocolFeeController(IVault(vaultAddress));
 
-        address deployedAddress = _create(abi.encode(vaultExtension, _authorizer, feeController), salt);
+        require(vaultCreationCodeHash == keccak256(vaultCreationCode), "VaultFactory: invalid vaultCreationCode");
```

Comment:
Are we still using `if - revert` with custom error? I remember we could have an issue with the compiler if we use require with custom error, but I'm not sure if we should use a require with a custom text. @EndymionJkb ?

Reply:
We use require with text for "preconditions" in tests, or things that shouldn't happen. In production code, for things that *could* happen, we should use custom errors.

I guess this is technically production code, though it's only used by us for deployments, not by regular users. We already have custom errors defined, so the question is whether this is a pre/post condition that should never be violated, or it's something that could reasonably happen. It shouldn't if you pass the right arguments, but it would be the only one, so I'd lean toward another custom error, though I don't feel strongly.

I tried using require with custom errors (which only works in 0.8.27). It increased the bytecode fairly dramatically (hundreds of bytes); not sure why. So I think we'll very likely stay with our current method, even if we go to 0.8.27. (We'll probably just stay on 0.8.26, though.)